### PR TITLE
support multi-module repos

### PR DESCRIPTION
**Pull request by Stebalien**
Run tests, build, vet, check, etc. for every module.

This also sets the default shell to "bash" (so this should work on windows).

### CODE REVIEW

**Review by mvdan — COMMENTED** [review]
I wrote this a while back after some trial and error across many repos at my previous job, if you want to reuse it:

```
go-modules() {
        find . \( -name vendor -o -name '[._].*' -o -name node_modules \) -prune -o -name go.mod -print | sed 's:/go.mod$::'
}
```

The main difference with yours is that it skips directories like Go, which might matter a lot if we have repositories that mix Go and e.g. JS. Running `find` through large directories like `node_modules` can take many seconds, which you would multiply by the number of steps here.

The line also trims the filename, so you can do:

```
for mod in $(go-modules); do
    (cd $mod && ...)
done
```

Also, it would be really useful to be able to declare this function/script once and reuse it. Be it a script like `for mod in $(./scripts/go-modules)`, or declaring a function in the job shell session, or a YAML anchor, etc. Maybe @marten-seemann has ideas.

One last thought - most of the error messages here assume a single module, so they would be confusing in multi-module repositories. I think they need a bit of work to disambiguate. For example:

> go.sum was added by go mod tidy (in what module?)

> foo.go: vet finds some mistake (in what module? paths are relative to the module root)

I see you actually dealt with this for staticcheck. We'd want to do the same for vet and the git line, I think.

**Comment by mvdan** [comment]
> skips directories like Go

I should clarify I also made it skip `node_modules`, which Go usually does not skip - precisely because we had a few large "full stack" repositories with both Go and JS. We used caching, so the `node_modules` directory was kept big even in CI.

**Comment by marten-seemann** [comment]
> Also, it would be really useful to be able to declare this function/script once and reuse it. Be it a script like for mod in $(./scripts/go-modules), or declaring a function in the job shell session, or a YAML anchor, etc.

I agree that defining a function would be nicer here. GitHub Actions doesn't support YAML anchors, so that's not an option.
`for mod in $(./scripts/go-modules)` would definitely work. Alternatively, we could have one step that determines the directories to run in, and use the result of that step in subsequent steps.
Not really sure what the cleanest solution would be.

> One last thought - most of the error messages here assume a single module, so they would be confusing in multi-module repositories. I think they need a bit of work to disambiguate. 

Poor man's solution:
```
for mod in $(go-modules); do
    echo "=== Running in $mod ==="
    (cd $mod && ...)
done
```

**Comment by mvdan** [comment]
I think I like "figure out the list once, and reuse it later" over anything else. It helps avoid an extra script in every repo, and if the checkout has tons of files, we also avoid running `find .` a handful of times.

**Comment by marten-seemann** [comment]
> I think I like "figure out the list once, and reuse it later" over anything else. It helps avoid an extra script in every repo

It helps avoid an extra script in every repo, but we will have to duplicate that step in both `go-test` and `go-check`. I really wish there was a way to share a GitHub Actions step between two workflows, but as far as I can see, there's no way to do that.
We *could* create our own GitHub Action for `go-modules()` though. Maybe that's the cleanest solution?

**Comment by marten-seemann** [comment]
Instead of looping over all directories we could also consider a matrix build. For every `go.mod` we'd spawn a new go-test and go-check. Conceptually, that's probably the cleanest solution.

There are 2 downsides to this:
1. This would lead to a massive number of builds in go-libp2p (every go-test already creates 6 jobs, for different OSes and Go versions), so we'd end up with ~50 builds for the 4 go.mods in that repository (multiply that by 2 since we run on `push` and on `pull_request`). But maybe that's ok for a multi-module repo?
2. For single-module repos, we'd still have a separate build step that determines the matrix. go-test and go-check would only start once that step has finished, so we'd slow down the build by maybe 20 seconds.

File: templates/.github/workflows/go-check.yml [hunk 1]
```diff
@@ -29,10 +32,14 @@ jobs:
...
         if: ${{ success() || failure() }} # run this step even if the previous one failed
-        run: go vet ./...
+        run: |
+          while read -r mod; do
+            ( cd "$(dirname "$mod")" && go vet ./... )
+          done < <(find -name go.mod)
       - name: staticcheck
         if: ${{ success() || failure() }} # run this step even if the previous one failed
         run: |
-          set -o pipefail
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply templates/.github/workflows/go-check.yml:36]: It's enabled by default for bash.

File: templates/.github/workflows/go-test.yml [hunk 2]
```diff
@@ -19,15 +22,24 @@ jobs:
           go version
           go env
       - name: Run tests
-        run: go test -v -coverprofile coverage.txt ./...
+        run: |
+          while read -r mod; do
+            ( cd "$(dirname "$mod")" && go test -v -coverprofile coverage.txt ./... )
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply templates/.github/workflows/go-test.yml:27]: As far as I can tell, yes. Codecov will look for all matching files in all directories.

**Comment by Stebalien** [comment]
All these approaches sound decent. I'd prefer not launch _too_ many actions, so triggering workflows is probably not an option (while I agree it's the "cleanest").

A "run this step in all go modules" action is probably the best tradeoff for now, as far as I can tell.

**Comment by Stebalien** [comment]
Specifically:

```
- with: protocollabs/run-in-all-go-modules
  run: |
    command 1
    command 2
```

But really, any of the options are probably fine.

**Comment by marten-seemann** [comment]
The `|` is just yaml syntax for multi-line commands, so that shouldn't be a problem.

After a bit of research, the syntax would be:
```yaml
- uses: protocollabs/run-in-all-go-modules
  with: 
     run: |
       command 1
       command 2
```

**Comment by Stebalien** [comment]
Maybe https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idoutputs is relevant?

**Pull request closed without merging**